File: arrays/accumulate.skeleton.js
// let example = [10, 20, 30, -1];
//
// console.log('array: ', example);
// console.log('sum:   ', sum(example));
// console.log('product: ', product(example));
// console.log('largest: ', largest(example));
// console.log('evens:   ', selectEvens(example));


function template(array) {
  let resultSoFar = _____;

  for (let value of array) {
    resultSoFar = _____;
  }

  return resultSoFar;
}

// Accumulator pattern
// Find-the-best pattern

function add(x, y) {
  return x + y;
}

// sum([]) === 0
function sum(array) {
  let resultSoFar = 0;

  for (let value of array) {
    resultSoFar = resultSoFar + value;
  }

  return resultSoFar;
}

// product([10, 10, 5]) === 10 * 10 * 5 === 500
// product([-1, 4, 200]) === -1 * 4 * 200 === -800
// product([]) === 1

function multiply(x, y) {
  return x * y;
}

function product(array) {
  let resultSoFar = _____;

  for (let value of array) {
    resultSoFar = _____;
  }

  return resultSoFar;
}

function larger(x, y) {
  if (x > y) {
    return x;
  } else {
    return y;
  }
}

function largest(array) {
  let resultSoFar = _____;

  for (let value of array) {
    resultSoFar = _____;
  }

  return resultSoFar;
}

function appendIfEven(array, num) {
  if (num % 2 === 0) {
    return array.concat(num);
  } else {
    return array;
  }
}

function selectEvens(array) {
  let resultSoFar = [];

  for (let value of array) {
    resultSoFar = appendIfEven(results, value);
  }

  return resultSoFar;
}

function applyAndAppend(array, item, fn) {
  return array.concat(fn(item));
}

function map(array, fn) {
  let resultSoFar = [];

  for (let value of array) {
    resultSoFar = applyAndAppend(array, value, fn);
  }

  return resultSoFar;
}

function accumulate(array, operation, initial) {
  let resultSoFar = initial;

  for (let item of array) {
    // console.log('resultSoFar:', resultSoFar);
    // console.log('item:', item);
    resultSoFar = operation(resultSoFar, item);
  }

  return resultSoFar;
}

// map([10, 20, 30], someFunc)
//   === [someFunc(10), someFunc(20), someFunc(30)]'


function mapAcc(array, someFunc) {
  return accumulate(
    array,
    (resultSoFar, item) => resultSoFar.concat(someFunc(item)),
    []
  );
}

function productAcc(array) {
  return accumulate(array, multiply, 1);
}


function largestAcc(array) {
  return accumulate(array, larger, -Infinity);
}

function sumAcc(array) {
  // array.reduce((x, y) => x + y, 0);
  return accumulate(array, (x, y) => x + y, 0);
}

// returns true if condition returns true for all
// elements of array
function everyAcc(array, condition) {
  return accumulate(
    array,
    (result, item) => result && condition(item),
    true
  );
}

function anyAcc(array, condition) {
  return accumulate(
    array,
    (result, item) => result || condition(item),
    false
  );
}

function countItemAcc(array, itemToCount) {
  return accumulate(
    array,
    (result, item) => item === itemToCount ? result + 1 : result,
    0
  );
}

function countItemsAcc(array) {
  return accumulate(
    array,
    (results, item) => results.set(results.has(item) ? results.get(item) + 1 : 1),
    new Map()
  );
}

File: arrays/accumulate.skeleton.py
# example = [10, 20, 30, -1]
#
# print('array: ', example);
# print('sum:   ', sum_list(example));
# print('product: ', product(example));
# print('largest: ', largest(example));
# print('evens:   ', select_evens(example));


import operator


def template(array):
    result_so_far = _____

    for value in array:
        result_so_far = _____

    return result_so_far

# Accumulator pattern

# Find-the-best pattern


def add(x, y):
    return x + y

# sum([]) === 0


def sum_list(array):
    result_so_far = 0

    for value in array:
        result_so_far = result_so_far + value

    return result_so_far


# product([10, 10, 5]) === 10  10  5 === 500

# product([-1, 4, 200]) === -1  4  200 === -800

# product([]) === 1

def multiply(x, y):
    return x * y


def product(array):
    result_so_far = _____

    for value in array:
        result_so_far = _____

    return result_so_far


def larger(x, y):
    if x > y:
        return x
    else:
        return y


def largest(array):
    result_so_far = _____

    for value in array:
        result_so_far = _____

    return result_so_far


def multiply(x, y):
    return x * y


def product(array):
    result_so_far = _____

    for value in array:
        result_so_far = _____

    return result_so_far


def append_if_even(array, num):
    if num % 2 == 0:
        return array + [num]
    else:
        return array


def select_evens(array):
    result_so_far = []

    for value in array:
        result_so_far = append_if_even(result_so_far, value)

    return result_so_far


def apply_and_append(array, item, fn):
    return array + [fn(item)]


def map(array, fn):
    result_so_far = []

    for value in array:
        result_so_far = apply_and_append(result_so_far, value, fn)

    return result_so_far


def accumulate(array, operation, initial):
    result_so_far = initial

    for item in array:
        # print('result_so_far:', result_so_far)
        # print('item:', item)
        result_so_far = operation(result_so_far, item)

    return result_so_far

# map([10, 20, 30], some_func)
#   === [some_func(10), some_func(20), some_func(30)]


def map_acc(array, some_func):
    return accumulate(
        array,
        lambda result_so_far, item: result_so_far + [some_func(item)],
        []
    )


def product_acc(array):
    return accumulate(array, operator.mul, 1)


def largest_acc(array):
    return accumulate(array, larger, -float('inf'))


def sum_acc(array):
    # array.reduce((x, y) => x + y, 0)
    return accumulate(array, lambda x, y: x + y, 0)

# returns true if condition returns true for all
# elements of array


def every_acc(array, condition):
    return accumulate(
        array,
        lambda result, item: result and condition(item),
        True
    )


def any_acc(array, condition):
    return accumulate(
        array,
        lambda result, item: result or condition(item),
        False
    )


def count_item_acc(array, item_to_count):
    return accumulate(
        array,
        lambda result, item: result + 1 if item == item_to_count else result,
        0
    )


def count_items_acc(array):
    return accumulate(
        array,
        lambda results, item: results.update(
            {item: results.get(item, 0) + 1}) or results,
        {}
    )
